texar/torch/modules/decoders/decoder_base.py

```diff
@@ -405,8 +406,22 @@ def dynamic_decode(self, helper: Helper, inputs: Optional[torch.Tensor],
 
         while (not torch.all(finished).item() and
                (max_decoding_length is None or time < max_decoding_length)):
-            (next_outputs, decoder_state, next_inputs,
-             decoder_finished) = self.step(helper, time, step_inputs, state)
+
+            next_outputs, decoder_state = \
+                self.compute_output_state(helper, time, step_inputs, state)
+
+            if max_decoding_length is not None and \
+                    time + 1 == max_decoding_length:
+                # Maximum decoding length reached, mark all batches as finished.
+                # This requires special handling because performing lookup on
+                # position embeddings with `time + 1` may result in IndexError.
+                decoder_finished = torch.tensor(1, dtype=torch_bool)
```

Comment:
This should probably be put on the same device as other tensors. Have you tested the code on GPU?

Reply:
good point. I think it should have the same `device` as that of `finished`.